src/mame/sony_news/cxd8452aq.cpp

```diff
@@ -0,0 +1,294 @@
+	static constexpr uint32_t INT_CLR_MASK = 0xf0;
+	static constexpr uint32_t RX_DMA_COMPLETE = 0x40;
+	static constexpr uint32_t TX_DMA_COMPLETE = 0x20;
+	static constexpr uint32_t EXT_INT = 0x1;
+
+	// count register constants
+	static constexpr uint32_t DMA_START = 0x80000000;
+
+	// DMA update timer
+	// TODO: Actual frequency, since we don't have DRQ to implictly rate limit
+	//       Might be the APbus frequency.
+	static constexpr int DMA_TIMER = 100;
+}
+
+cxd8452aq_device::cxd8452aq_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, CXD8452AQ, tag, owner, clock),
+	  device_memory_interface(mconfig, *this),
+	  main_bus_config("main_bus", ENDIANNESS_BIG, 32, 32, 0),
+	  sonic_config("sonic", ENDIANNESS_BIG, 32, 32, 0, address_map_constructor(FUNC(cxd8452aq_device::sonic_bus_map), this)),
+	  m_irq_handler(*this),
+	  m_apbus_virt_to_phys_callback(*this),
+	  m_bus(*this, finder_base::DUMMY_TAG, -1, 64) {}
+
+void cxd8452aq_device::map(address_map &map)
+{
+	map(0x00, 0x03).rw(FUNC(cxd8452aq_device::control_r), FUNC(cxd8452aq_device::control_w));
+	map(0x04, 0x07).rw(FUNC(cxd8452aq_device::config_r), FUNC(cxd8452aq_device::config_w));
+	map(0x08, 0x0b).r(FUNC(cxd8452aq_device::revision_r));
+	map(0x0c, 0x0f).rw(FUNC(cxd8452aq_device::rx_sonic_addr_r), FUNC(cxd8452aq_device::rx_sonic_addr_w));
+	map(0x10, 0x13).rw(FUNC(cxd8452aq_device::rx_host_addr_r), FUNC(cxd8452aq_device::rx_host_addr_w));
+	map(0x14, 0x17).rw(FUNC(cxd8452aq_device::rx_count_r), FUNC(cxd8452aq_device::rx_count_w));
+	map(0x18, 0x1b).rw(FUNC(cxd8452aq_device::tx_sonic_addr_r), FUNC(cxd8452aq_device::tx_sonic_addr_w));
+	map(0x1c, 0x1f).rw(FUNC(cxd8452aq_device::tx_host_addr_r), FUNC(cxd8452aq_device::tx_host_addr_w));
+	map(0x20, 0x23).rw(FUNC(cxd8452aq_device::tx_count_r), FUNC(cxd8452aq_device::tx_count_w));
+}
+
+void cxd8452aq_device::sonic_bus_map(address_map &map)
+{
+	map(0x0, 0xfffff).mirror(0xfff00000).rw(FUNC(cxd8452aq_device::sonic_r), FUNC(cxd8452aq_device::sonic_w));
+}
+
+void cxd8452aq_device::device_start()
+{
+	m_bus->cache(m_net_cache);
+	m_irq_handler.resolve_safe();
+	m_apbus_virt_to_phys_callback.resolve();
+	m_irq_check = machine().scheduler().timer_alloc(timer_expired_delegate(FUNC(cxd8452aq_device::irq_check), this));
+	m_dma_check = machine().scheduler().timer_alloc(timer_expired_delegate(FUNC(cxd8452aq_device::dma_check), this));
+
+	save_item(STRUCT_MEMBER(m_sonic3_reg, control));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, config));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, revision));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, rx_sonic_address));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, rx_host_address));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, rx_count));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, tx_sonic_address));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, tx_host_address));
+	save_item(STRUCT_MEMBER(m_sonic3_reg, tx_count));
```

Comment:
You only need `STRUCT_MEMBER` when you have arrays of structures.  For a single structure, you can just `save_item(NAME(m_sonic3_reg.control));` etc.

Reply:
Fixed, thanks!